src/xdp/apikernel.c

```diff
@@ -0,0 +1,539 @@
+//
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+//
+
+#include "precomp.h"
+#include "apikernel.tmh"
+
+typedef struct _XDPAPI_PROVIDER {
+    NPI_PROVIDER_CHARACTERISTICS Characteristics;
+    HANDLE NmrProviderHandle;
+} XDPAPI_PROVIDER;
+
+typedef struct _XDPAPI_CLIENT {
+    HANDLE NmrBindingHandle;
+    GUID ClientModuleId;
+    const XDP_API_CLIENT_DISPATCH *ClientDispatch;
+    const VOID *ClientBindingContext;
+} XDPAPI_CLIENT;
+
+typedef struct _XDP_API_ROUTINE {
+    _Null_terminated_ const CHAR *RoutineName;
+    VOID *Routine;
+} XDP_API_ROUTINE;
+
+static XDPAPI_PROVIDER XdpApiProvider;
+static const NPI_MODULEID NPI_XDP_MODULEID = {
+    sizeof(NPI_MODULEID),
+    MIT_GUID,
+    { 0x521421C2, 0x539B, 0x46DF, { 0xA1, 0x43, 0x8D, 0xAC, 0x88, 0xD0, 0xC1, 0x55 } }
+};
+
+//
+// API routines.
+//
+static XDP_GET_ROUTINE_FN XdpApiKernelXdpGetRoutine;
```

Comment:
these names are really long. let's call them XdpGetRoutineKernel instead, i.e. use a one-word suffix instead of three-word prefix.

Reply:
actually, why not just call it XdpGetRoutine?

Reply:
@nigriMSFT any idea?

Reply:
I don't have a preference on the naming, but agree that they were getting long